src/sql/src/plan/query.rs

```diff
@@ -58,6 +61,147 @@ use crate::plan::statement::StatementContext;
 use crate::plan::transform_ast;
 use crate::plan::typeconv::{self, CastContext};
 
+// Aug is the type variable assigned to an AST that has already been
+// name-resolved. An AST in this state has global IDs populated next to table
+// names, and local IDs assigned to CTE definitions and references.
+#[derive(Debug, PartialEq, Eq, Hash, Copy, Clone, Default)]
+pub struct Aug;
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub struct ResolvedTableName {
+    id: Id,
+    raw_name: ObjectName,
+}
+
+impl AstDisplay for ResolvedTableName {
+    fn fmt(&self, f: &mut AstFormatter) {
+        f.write_str(format!("[{}: ", self.id));
+        separated(&self.raw_name.0, ".").fmt(f);
+        f.write_str("]");
+    }
+}
+
+impl std::fmt::Display for ResolvedTableName {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        f.write_str(self.to_ast_string().as_str())
+    }
+}
+
+impl AstInfo for Aug {
+    type Table = ResolvedTableName;
+    type Id = Id;
+}
+
+#[derive(Debug)]
+struct NameResolver<'a, 'b> {
+    qcx: &'a mut QueryContext<'b>,
+    status: Result<(), anyhow::Error>,
+}
+
+impl<'a, 'b> Fold<Raw, Aug> for NameResolver<'a, 'b> {
+    fn fold_query(&mut self, q: Query<Raw>) -> Query<Aug> {
+        // Retain the old values of various CTE names so that we can restore them after we're done
+        // planning this SELECT.
+        let mut old_cte_values = Vec::new();
+        // A single WITH block cannot use the same name multiple times.
+        let mut used_names = HashSet::new();
+        let mut ctes = Vec::new();
+        for cte in q.ctes {
+            let cte_name = normalize::ident(cte.alias.name.clone());
+
+            if used_names.contains(&cte_name) {
+                self.status = Err(anyhow!(
+                    "WITH query name \"{}\" specified more than once",
+                    cte_name
+                ));
+            }
+            used_names.insert(cte_name.clone());
+
+            let id = LocalId::new(self.qcx.ctes.len() as u64);
+            ctes.push(Cte {
+                alias: cte.alias,
+                id: Id::Local(id),
+                query: self.fold_query(cte.query),
+            });
+            let old_val = self.qcx.ctes.insert(cte_name.clone(), id);
+            old_cte_values.push((cte_name, old_val));
+        }
+        let result = Query {
+            ctes,
+            body: self.fold_set_expr(q.body),
+            limit: q.limit.map(|l| self.fold_limit(l)),
+            offset: q.offset.map(|l| self.fold_expr(l)),
+            order_by: q
+                .order_by
+                .into_iter()
+                .map(|c| self.fold_order_by_expr(c))
+                .collect(),
+        };
```

Comment:
I _think_ you should be able to use `sql_parser::ast::fold::fold_query` here, yeah?